core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,170 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+#define KOKKOS_IMPL_HOST_SHARED_PTR_HPP
+
+#include <Kokkos_Macros.hpp>
+#include <Kokkos_Atomic.hpp>
+
+#include <functional>
+
+namespace Kokkos {
+namespace Impl {
+
+template <typename T>
+class HostSharedPtr {
+ public:
+  using element_type = T;
+
+  explicit HostSharedPtr(T* element_ptr = nullptr)
+      : HostSharedPtr(element_ptr, [](T* const t) { delete t; }) {}
+
+  template <class Deleter>
+  HostSharedPtr(T* element_ptr, const Deleter& deleter)
+      : m_element_ptr(element_ptr), m_control(nullptr) {
+    if (element_ptr) {
+      try {
+        m_control = new Control{deleter, 1};
+      } catch (...) {
+        deleter(element_ptr);
```

Comment:
Minor/non-blocking: since `deleter` can be `nullptr` and still represent a valid construction for `std::function`, this should be:

```suggestion
        if(deleter) { deleter(element_ptr); }
```

Reply:
I'm not quite sure if passing `nullptr` as deleter should be considered legal but adding the check probably doesn't hurt.

Reply:
@dhollman does that mean you'd want to guard https://github.com/kokkos/kokkos/blob/e67449321ca5a65e61048fc0b0e29a3977f9b676/core/src/impl/Kokkos_HostSharedPtr.hpp#L151 as well

Reply:
Honestly I'd rather do `KOKKOS_EXPECT(bool(m_control->m_deleter))` in `Control::Control(Deleter const&)`

Reply:
Uggghh it does not have constructors.  Anyway I mean make it a precondition.

Reply:
I think a precondition is fine for the interface.

Since `nullptr` is not invocable, The only way to accidentally pass a null pointer to the deleter is something like:

```
using F = void (*)(T*);
HostSharedPtr<T>(GetT(), F());
```

Internally, I agree that if `m_control` is set, `m_control->m_deleter` should be invocable (class invariant).

Reply:
I still think that `nullptr` is clearer than `[](T*){}` for a do-nothing deleter, and it has the advantage of removing a type from the binary (the lambda) and a jump from the destructor of the version that references the singleton (replaced with a branch, which is likely a tiny bit faster). But like I said, it's a minor issue, and if I'm in the minority then it's fine as is.

Reply:
Just putting this here to support my reasoning for future reference: https://quick-bench.com/q/JXWqAeOJ-rszy3b4y1SLI-LQ4CY